File: react-frontend/src/types/models/products/Product.ts
import type { Category } from "./Category";

export interface ImagePayload {
  id: number,
  imageUrl: string,
  displayOrder: number,
  isMain: boolean,
  productId: number | null,
  variantId: number | null,
}

export interface ProductVariantOptionValue {
  optionValueId: number;
  optionName: string;
  value: string;
}

export interface ProductVariant {
  id: number;
  variantName: string;
  sku: string;
  price: number;
  originalPrice: number;
  stock: number;
  productId: number;
  images: ImagePayload[];
  optionValues: ProductVariantOptionValue[];
}

export interface ProductOptionValue {
  optionValueId: number;
  value: string;
}

export interface ProductOption {
  optionId: number;
  optionName: string;
  optionValues: ProductOptionValue[];
}

export interface Product {
  id: number | string;
  name: string;
  slug: string;
  basePrice: number;
  description?: string;
  shortDescription?: string;
  thumbnailUrl: string;
  categoryId?: number | string;
  category?: Category;
  options?: ProductOption[];
  selectedOptionValueIds?: number[];
  stock: number;
  variants: ProductVariant[];
  createdAt: string;
  updatedAt: string;
}

File: dotnet_api/src/Entities/Product.cs
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
using WebApp_API.Enums;

namespace WebApp_API.Entities
{
    public class Product
    {
        [Key] public int Id { get; set; }
        [Required, MaxLength(300)] public required string Name { get; set; }
        [Required, MaxLength(300)] public required string Slug { get; set; }

        [MaxLength(1000)] public string? ShortDescription { get; set; }
        public string? Description { get; set; }
        [Column(TypeName = "decimal(18,2)")] public decimal BasePrice { get; set; }
        [MaxLength(1000)] public string? ThumbnailUrl { get; set; }

        public required int CategoryId { get; set; }
        [ForeignKey("CategoryId")] public Category? Category { get; set; }

        public DateTime CreatedAt { get; set; }
        public DateTime UpdatedAt { get; set; }

        public ProductStatus Status { get; set; } = ProductStatus.Draft;

        public bool IsDeleted { get; set; }
        public DateTime? DeletedAt { get; set; }

        public ICollection<ProductVariant> Variants { get; set; } = new List<ProductVariant>();
        public ICollection<ProductImage> Images { get; set; } = new List<ProductImage>();
        public ICollection<ProductFilter> ProductFilters { get; set; } = new List<ProductFilter>();
        public ICollection<OrderItem> OrderItems { get; set; } = new List<OrderItem>();
    }
}